presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -134,94 +194,377 @@ public boolean needsInput()
+
+            if (currentRowSpilled) {
+                savedRows.merge(
+                        currentRowPartition,
+                        new SavedRow(currentPage, currentPosition, joinPositionWithinPartition, currentProbePositionProducedRow, joinSourcePositions),
+                        (oldValue, newValue) -> {
+                            throw new IllegalStateException(format("How on earth this could happen that partition %s is spilled in the middle of processing twice?", currentRowPartition));
+                        });
+                Page unprocessed = pageTail(currentPage, currentPosition + 1);
+                addInput(unprocessed, spillInfoSnapshot);
+            }
+            else {
+                Page remaining = pageTail(currentPage, currentPosition);
+                restoreProbe(remaining, currentJoinPosition, currentProbePositionProducedRow, joinSourcePositions, spillInfoSnapshot);
+            }
+        }
     }
 
-    private Page getOutput(LookupSource lookupSource)
+    private void processProbe(LookupSource lookupSource)
     {
-        // join probe page with the lookup source
+        verify(probe != null);
+
         DriverYieldSignal yieldSignal = operatorContext.getDriverContext().getYieldSignal();
-        if (probe != null) {
-            while (!yieldSignal.isSet()) {
-                if (probe.getPosition() >= 0) {
-                    if (!joinCurrentPosition(lookupSource, yieldSignal)) {
+        while (!yieldSignal.isSet()) {
+            if (probe.getPosition() >= 0) {
+                if (!joinCurrentPosition(lookupSource, yieldSignal)) {
+                    break;
+                }
+                if (!currentProbePositionProducedRow) {
+                    currentProbePositionProducedRow = true;
+                    if (!outerJoinCurrentPosition(lookupSource)) {
                         break;
                     }
-                    if (!currentProbePositionProducedRow) {
-                        currentProbePositionProducedRow = true;
-                        if (!outerJoinCurrentPosition(lookupSource)) {
-                            break;
-                        }
-                    }
-                }
-                currentProbePositionProducedRow = false;
-                if (!advanceProbePosition(lookupSource)) {
-                    break;
                 }
-                statisticsCounter.recordProbe(joinSourcePositions);
-                joinSourcePositions = 0;
             }
+            currentProbePositionProducedRow = false;
+            if (!advanceProbePosition(lookupSource)) {
+                break;
+            }
+            statisticsCounter.recordProbe(joinSourcePositions);
+            joinSourcePositions = 0;
         }
+    }
+
+    private void restoreProbe(Page probePage, long joinPosition, boolean currentProbePositionProducedRow, int joinSourcePositions, SpillInfoSnapshot spillInfoSnapshot)
```

Comment:
I think it might be easier to follow, if this took a `SavedRow` even for the case in `processProbe`

Reply:
I wanted to avoid `SavedRow` instantiation since it involves page compaction.